openmdao/visualization/meta_model_viewer/meta_model_visualization.py

```diff
@@ -0,0 +1,551 @@
+        
+        z_data = self.Z[alt_index].flatten()
+        self.source.add(z_data, 'bot_slice')
+
+        x = self.slider_source.data[self.x_input.value]
+        y = self.source.data['bot_slice']
+
+        s2 = figure(plot_width=550, plot_height=200, x_range=(min(self.x_data),max(self.x_data)), y_range=(min(y), max(y)), 
+        title="{} vs {}".format(self.x_input.value, self.output_select.value), tools="")
+        s2.xaxis.axis_label = self.x_input.value
+        s2.yaxis.axis_label = self.output_select.value
+        s2.line(x, y)
+
+        data = self.training_points()
+        horiz_color = np.zeros((len(data), 1))
+        for i,info in enumerate(data):
+            alpha = np.abs(info[1] - self.y_value) / self.limit_range[self.y_index]
+            if alpha < self.dist_range:
+                horiz_color[i, -1] = (1 - alpha / self.dist_range) * info[-1]
+        
+        color = np.column_stack((data[:,-4:-1] - 1, horiz_color))
+        alphas = [0 if math.isnan(x) else x for x in color[:, 3]]
+        s2.scatter(x=data[:, 0], y=data[:, 3], line_color=None, fill_color='#000000', fill_alpha=alphas)
+
+        self.bot_plot_source.data = dict(bot_slice_x=self.x_data, bot_slice_y=np.repeat(self.y_value, self.resolution))
+        self.contour_plot.line('bot_slice_x', 'bot_slice_y', source=self.bot_plot_source, color='black', line_width=2)
+
+        return s2
+
+    def update_all_plots(self):
+        self.layout.children[0] = self.contour_data()
+        self.layout.children[1] = self.right_plot()
+        self.layout2.children[0] = self.bot_plot()
+
+    def update_subplots(self):
+        self.layout.children[1] = self.right_plot()
+        self.layout2.children[0] = self.bot_plot()
+
+
+    # Event handler functions
+    def update(self, attr, old, new):
+        self.update_all_plots()
+
+    def scatter_plots_update(self, attr, old, new):
+        self.update_subplots()
+
+    def scatter_input(self, attr, old, new):
+        self.dist_range = float(new)
+        self.update_all_plots()
+
+    def input_dropdown_checks(self,x,y):
+        # Checks to see if x and y inputs are equal to each other
+        if x == y:
+            return False
+        else:
+            return True
+    
+    def x_input_update(self, attr, old, new):
+        if self.input_dropdown_checks(new, self.y_input.value) == False:
+            raise ValueError("Inputs should not equal each other")
+        else:
+            self.update_all_plots()
+
+    def y_input_update(self, attr, old, new):
+        if self.input_dropdown_checks(self.x_input.value, new) == False:
+            raise ValueError("Inputs should not equal each other")
+        else: 
+            self.update_all_plots()
+
+    def output_value_update(self, attr, old, new):
+        self.output_variable = self.output_list.index(new)
+        self.update_all_plots()
+
+    def training_points(self):
```

Comment:
I would suggest that you write a unit test for this function specifically, since it seems easily isolatable functionality